Implement ConstConfigIOGroup

avilcheslopez opened the pull request:
Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Relates to #2616 
- Fixes #2619, #2625, #2620, #2621, #2622, #2623, #2624, #2660

avilcheslopez committed:
Add JSON schema for ConstConfigIOGroup config file

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add initial ConstConfigIOGroup class

Note: initial stubbed out class (to enable writing unit tests).

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add convenience constructor in ConstConfigIOGroup

Add convenience constructor to enable unit testing.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Implement JSON config parsing in ConstConfigIOGroup

Also updated the JSON schema to make it simpler and more compact.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on service/json_schemas/const_config_io.schema.json:
```diff
@@ -0,0 +1,39 @@
+                "values": {
+                    "type": "array",
+                    "items": {
+                        "type": "number"
```
Consider making it `["number", "null"]` or `["number", "string"]` if you want to allow some parts of a domain to hold no value. I think the latter is more consistent with what we allow for policies. Bonus if you use a pattern-matched string type that accepts variations of `"nan"`.

avilcheslopez replied:
After reviewing with Brad and Lowren, they were suggesting to simply not allow nulls or nans (i.e. enforce all values are provided). If there's a valid use case to allow nulls/nans, let's revisit.

dannosliwcd commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+
+    ConstConfigIOGroup::ConstConfigIOGroup()
+    {
+        // TODO
```
I think this is sufficient as an empty constructor, since it just means we'll expose no signals, right?

avilcheslopez replied:
See my reply to Brad's comment.

dannosliwcd commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+        return M_PLUGIN_NAME;
+    }
+
+    std::unique_ptr<IOGroup> ConstConfigIOGroup::make_plugin(void)
+    {
+        return geopm::make_unique<ConstConfigIOGroup>();
+    }
```
How will this change for non-empty const configs? Maybe managed similar to other geopm config files?

avilcheslopez replied:
I think I'll be able to get away with still invoking the default constructor once I update it to load the config from a file (which could be provided via an environment variable according to the feature description).

dannosliwcd commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+                            std::to_string(batch_idx) + " out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        m_signal_ref_s signal = m_pushed_signals[batch_idx];
+        auto it = m_signal_available.begin();
+        std::advance(it, signal.signal_idx);
+        result = it->second.values[signal.domain_idx];
```
I think this only works if available signals never changes in between calls to this (which would potentially change the offsets as new keys are inserted/removed). Looks your current implementation does avoid that.

Should we enforce that by making `m_signal_available` a `const` member? You'd need to initialize it in the constructor instead of post-construction. E.g., `ConstConfigIOGroup(const std::string &config) : m_signal_available(something_that_returns_the_map(config))`

bgeltz replied:
I'm in favor of this suggestion.

avilcheslopez replied:
Yeah, I'm not fully convinced with my initial implementation. I'll revisit this soon. If I do stick with this implementation, I'll make sure to follow your suggestion of making ``m_signal_available`` a const member.

Krebrov001 committed:
WIP work on `ConstConfigIOGroupTest`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "ConstConfigIOGroup.hpp"
+
+#include <cmath>
+
+#include <iostream>
+#include <iterator>
+#include <sstream>
+
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Agg.hpp"
+#include "geopm_topo.h"
+#include "config.h"
```
```suggestion
#include "config.h"

#include "ConstConfigIOGroup.hpp"

#include <cmath>
#include <iostream>
#include <iterator>
#include <sstream>

#include "geopm_topo.h"
#include "geopm/Helper.hpp"
#include "geopm/PlatformTopo.hpp"
#include "geopm/Exception.hpp"
#include "geopm/Agg.hpp"
```

bgeltz replied:
Is Helper.hpp needed for anything?  I thought read_file() would be called somewhere, but I don't see it.

avilcheslopez replied:
Will reorder the includes, thanks!

The code is currently using ``make_unique()`` from ``Helper.hpp``, but yes: the intention is to add code to load the config from a file (i.e. ``read_file()`` will be called).

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+        ConstConfigIOGroup::M_SIGNAL_FIELDS = {
+            {"description", {Json::STRING, true}},
+            {"units", {Json::STRING, true}},
+            {"domain", {Json::STRING, true}},
+            {"aggregation", {Json::STRING, true}},
+            {"values", {Json::ARRAY, true}}
+        };
+
+    ConstConfigIOGroup::ConstConfigIOGroup()
+    {
```
I thought this would look something like:
```suggestion
    ConstConfigIOGroup::ConstConfigIOGroup()
        : ConstConfigIOGroup(M_DEFAULT_CONFIG_FILE)
    {
```

avilcheslopez replied:
Addressing both (this and Daniel's) comments here:

The feature description calls for accepting a config file path via an environment variable. So the default constructor (called by the ``make_plugin()`` method) will read the config file (perhaps a default path could be used if none provided) and then proceed with parsing. I provided the specialized constructor as a convenience for the unit tests. I might refactor this a bit if I can think of a better solution.

bgeltz replied:
I think we need to refactor the constructor a bit.  There are some code paths that we cannot test in CI the way it is.  I put a sketch together with a few TODOs.  It's here:
https://github.com/bgeltz/geopm/commit/f51e3e84f4c8d5f19b0c86c6c8c4005b3b470a0c
https://github.com/bgeltz/geopm/tree/issue-2619

With a refactor like I suggested, this is easier to test and maintain.  Take a look and let me know if you have questions.  The unit tests should create real/on-disk input files that are parsed here.  Both paths should be tested (env specified file path and default path).

Krebrov001 replied:
@bgeltz Do you want me to write the unit tests that create real/on-disk input files?

bgeltz replied:
Yes

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+
+    ConstConfigIOGroup::ConstConfigIOGroup(const std::string &config)
+    {
+        parse_config_json(config);
```
Where is the checking to see if the config file exists?  As it is, it seems if you pass a string to the parse method, and the file does not exist that this will throw, and construction will fail.

bgeltz replied:
Also, where is the file reading code?

avilcheslopez replied:
It's coming :). I first wanted to enable Konstantin to start writing unit tests.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+                                " unexpected propety: \"" + prop.first +
+                                "\"", GEOPM_ERROR_INVALID, __FILE__,
+                                __LINE__);
+            } else if (prop.second.type() != it->second.type) {
```
```suggestion
            }
            else if (prop.second.type() != it->second.type) {
```

avilcheslopez replied:
Thanks!

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,424 @@
+                                " incorrect type for property: \"" +
+                                prop.first + "\"", GEOPM_ERROR_INVALID,
+                                __FILE__, __LINE__);
+            } else {
```
```suggestion
            }
            else {
```

bgeltz commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,90 @@
+    EXPECT_EQ(iogroup.is_valid_control("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), false);
+}
+
+void ConstConfigIOGroupTest::TearDown()
+{
+}
```
Move this under SetUp

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,90 @@
+TEST_F(ConstConfigIOGroupTest, valid_json)
+{
+    std::string json_string = "{"
+    "    \"GPU_CORE_FREQUENCY\": {"
+    "        \"domain\": \"gpu\","
+    "        \"description\": \"Provides GPU core frequency\","
+    "        \"units\": \"hertz\","
+    "        \"aggregation\": \"sum\","
+    "        \"values\": { 1500, 1600, 1700 }"
```
Note: don't forget to push the fix for this: use square brackets for an array.

avilcheslopez replied:
Make sure to cover more interesting configs with multiple signals.

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,90 @@
+
+TEST_F(ConstConfigIOGroupTest, control_names)
+{
+    ConstConfigIOGroup iogroup("");
```
For this test, you'll need to pass a valid JSON string to ensure the class loads correctly, and then ensure the ``control_names()`` method returns an empty set. In other words, the behavior we're validating is: given a valid config, ConstConfigIOGroup always returns an empty set for the control names (as controls are not supported).

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,90 @@
+
+TEST_F(ConstConfigIOGroupTest, is_valid_control)
+{
+    ConstConfigIOGroup iogroup("");
```
Same comment as before.

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,90 @@
+    );
+}
+
+TEST_F(ConstConfigIOGroupTest, input_gibberish)
```
This is a good start for the negative tests. In case you don't have these in mind, make sure you cover cases like:
- Signals with missing required fields
- Fields with values of an incorrect type
- Fields with a value outside of their domain/enum
- Duplicate signals
- A signal with an empty "values" array

avilcheslopez reviewed (commented):
Tests are looking good! I'm sure you're already working on it, but just in case, make sure to cover all the other methods of the class.

If while writing a test you see the class is not behaving as you'd expect, please let me know, as this might be a bug I need to fix.

Krebrov001 committed:
Adding positive and negative ConstConfigIOGroup tests

- Testing all the query methods based on a valid inputs.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

avilcheslopez committed:
Load ConstConfigIOGroup config from file

- Load the config from a file specified by the user via an environment
  variable
- Handle case of duplicate signals defined in config
- Minor improvements and clean-up

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add chapter 7 man page for ConstConfigIOGroup

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

Krebrov001 committed:
Added new JSON input tests to `ConstConfigIOGroup`

 - `ConstConfigIOGroupTest.input_duplicate_signal`
 - `ConstConfigIOGroupTest.input_missing_properties`
 - `ConstConfigIOGroupTest.input_unexpected_properties`
 - `ConstConfigIOGroupTest.input_duplicate_properties`
 - `ConstConfigIOGroupTest.input_invalid_domain`
 - `ConstConfigIOGroupTest.input_invalid_units`
 - `ConstConfigIOGroupTest.input_invalid_aggregation`
 - `ConstConfigIOGroupTest.input_incorrect_type`
 - `ConstConfigIOGroupTest.input_array_value_type`
 - `ConstConfigIOGroupTest.input_array_value_empty`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,466 @@
+        ConstConfigIOGroup iogroup(json_string),
+        GEOPM_ERROR_INVALID,
+        "ConstConfigIOGroup::parse_config_json(): "
+        "missing properties for signal \"" "GPU_CORE_FREQUENCY" "\": "
```
Question: does the error message include the missing properties? If not, that's a bug in ConstConfigIOGroup.

Krebrov001 committed:
Add two new input tests for `ConstConfigIOGroup`

 - `ConstConfigIOGroupTest.input_empty_json`
 - `ConstConfigIOGroupTest.input_capital_properties`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add tests to check for empty fields.

 - `ConstConfigIOGroupTest.input_empty_domain`
 - `ConstConfigIOGroupTest.input_empty_description`
 - `ConstConfigIOGroupTest.input_empty_units`
 - `ConstConfigIOGroupTest.input_empty_aggregation`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+TEST_F(ConstConfigIOGroupTest, input_empty_json)
+{
+    ConstConfigIOGroup iogroup("{}");
+    EXPECT_EQ(iogroup.signal_names(), std::set<std::string>{});
```
Check ``control_names()`` also returns an empty set.

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+
+    EXPECT_EQ(iogroup.control_names(), std::set<std::string>{});
+    EXPECT_EQ(iogroup.is_valid_signal("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), false);
+    EXPECT_EQ(iogroup.is_valid_control("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), false);
```
Also check: ``is_valid_control("CONST_CONFIG::GPU_CORE_FREQUENCY")``

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+    EXPECT_EQ(iogroup.is_valid_signal("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), false);
+    EXPECT_EQ(iogroup.is_valid_control("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), false);
+    EXPECT_EQ(iogroup.signal_domain_type("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), GEOPM_DOMAIN_INVALID);
+    EXPECT_EQ(iogroup.control_domain_type("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"), GEOPM_DOMAIN_INVALID);
```
Also check: ``control_domain_type("CONST_CONFIG::GPU_CORE_FREQUENCY")``

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+        "3" " out of range."
+    );
+    GEOPM_EXPECT_THROW_MESSAGE(
+        iogroup.push_control("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO", 0, 0),
```
Also check: ``push_control("CONST_CONFIG::GPU_CORE_FREQUENCY")``

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+        "3" " out of range."
+    );
+    GEOPM_EXPECT_THROW_MESSAGE(
+        iogroup.write_control("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO", 0, 0, 3.14),
```
Also check: ``write_control("CONST_CONFIG::GPU_CORE_FREQUENCY")``

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+        " not valid for ConstConfigIOGroup"
+    );
+    GEOPM_EXPECT_THROW_MESSAGE(
+        iogroup.control_description("CONST_CONFIG::UNCORE_RATIO_LIMIT:MIN_RATIO"),
```
Also check: ``control_description("CONST_CONFIG::GPU_CORE_FREQUENCY")``

avilcheslopez commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,565 @@
+
+TEST_F(ConstConfigIOGroupTest, input_duplicate_signal)
+{
+    /* When one signal is provided more than once,
+       apparently the latter entry overwrites the data of the former. */
```
Suggested change in comment: When one signal is provided more than once, the latest entry will take precedence. This is due to how the underlying JSON library handles duplicate entries.

avilcheslopez committed:
Load default ConstConfigIOGroup config file

When unable to load config file provided by the user, fall back to
loading a default config file.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

Krebrov001 committed:
Addressing review updates

 - Check that upon missing properties, the exception displays their names.
 - Check that the `control_names` are empty whenever the `signal_names` get queried.
 - Check that `ConstConfigIOGroup` doesn't mixup the `signal_names` with the `control_names`.
 - Edit the comment for the duplicate input signals to clarify default behavior of the JSON library.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on copying_headers/MANIFEST.EXEMPT:
```diff
@@ -128,6 +130,7 @@ service/docs/source/geopm_imbalancer.3.rst
 service/docs/source/geopm_pio.3.rst
 service/docs/source/geopm_pio.7.rst
+service/docs/source/geopm_pio_const_config.7.rst
 service/docs/source/geopm_pio_cnl.7.rst
```
```suggestion
service/docs/source/geopm_pio_cnl.7.rst
service/docs/source/geopm_pio_const_config.7.rst
```

bgeltz commented on service/docs/source/geopm_pio_const_config.7.rst:
```diff
@@ -0,0 +1,80 @@
+-----------
+
+ConstConfigIOGroup implements the
+:doc:`geopm::IOGroup(3) <GEOPM_CXX_MAN_IOGroup.3>` interface to provide a
+predefined set of signals by the user with constant values. This is mainly
+useful to provide configuration information needed by some agents that's
+constant in nature, for example, system characterization information.
```
```suggestion
:doc:`geopm::IOGroup(3) <GEOPM_CXX_MAN_IOGroup.3>` interface to provide a
predefined set of signals with constant values to the user. This is mainly
useful to provide configuration information needed by some agents (e.g., system characterization information).
```

avilcheslopez replied:
Just to clarify, the reason for saying the signals are provided by the user is to explain that this is a user-defined set of signals. Maybe there's a better way of wording that. Maybe it should be mentioned that the administrator is meant to generate the configuration.

The concern I have with the suggestion is that technically the signals are provided to the IOGroup, not the user (unless I'm misunderstanding what you meant).

bgeltz replied:
I think we need to be careful how we document this.  This IOGroup allows for signals to be either: user-defined (i.e., the user wrote a config file and set GEOPM_CONST_CONFIG_PATH to point to the file they wrote) or admin defined (i.e., GEOPM_CONST_CONFIG_PATH unset but the default file exists and has entries), or neither, or both user and admin defined.

It should be clear in the docu that:
   a. This could be leveraged by the user, the sysadmin, neither, or both.
   b. How we decide who wins if both files exist, and the env is set properly, or if only one of the files exists, etc.

avilcheslopez replied:
Yes. That needs to be addressed (realized later that the documentation needed to be updated to reflect the latest changes in the code). Just wasn't sure about your original suggestion about a wording change.

bgeltz commented on service/json_schemas/const_config_io.schema.json:
```diff
@@ -0,0 +1,39 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "id": "https://geopm.github.io/const_config_io.schema.json",
```
Reminder to update the website to include this after this PR is merged.

Krebrov001 replied:
It's not clear to me what needs to be done here.

bgeltz replied:
When this PR is merged, this new json file will need to be added to the repo that drives the website.

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+        public:
+            ConstConfigIOGroup();
+            ConstConfigIOGroup(const std::string &config);
+            /// @return the list of signal names provided by this IOGroup.
```
```suggestion
```
It looks like most of these docstrings were boiler plate comments from either the CNLIOgroup or the CpuinfoIOGroup.  Most other IOGroups do not add these docstrings.  IMO they're not adding any benefit here.

If there are any special details about this specific class, please preserve them.  Otherwise I'd remove all the docstrings that were copy/pasted from the mentioned IOGroups.

avilcheslopez replied:
Will do! I'll keep the ones that document behavior that's particular to this IOGroup.

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+                bool required;
+            };
+
+            struct m_signal_info_s {
+                m_signal_info_s(
+                    int units,
+                    int domain,
+                    std::function<double(const std::vector<double> &)> agg,
+                    std::string desc,
+                    std::vector<double> values):
+
+                    units(units),
+                    domain(domain),
+                    agg_function(agg),
+                    description(desc),
+                    values(values)
+                {}
+
+                int units;
+                int domain;
+                std::function<double(const std::vector<double> &)> agg_function;
+                std::string description;
+                std::vector<double> values;
+            };
```
I think this can be simplified a bit.  I think there are some redundant/unused fields with the current implementation.  Will this work?

```suggestion            
            struct m_signal_info_s
            {                
                int domain;
                int units;
                std::function<double(const std::vector<double> &)> agg_function;
                std::string description;
                std::vector<double> values;
            };
```

avilcheslopez replied:
Wait, it seems your suggestion is to get rid of the constructor (i.e. I'm not seeing changes related to potentially redundant fields). The constructor is needed as a convenience when calling ``std::make_shared<m_signal_info_s>(...)``.

bgeltz replied:
Is there a reason that ```m_signal_available``` is defined that way (having a shared pointer to the signal info struct?  In all other IOGroups I looked at, the definition is just:
```
std::map<std::string, m_signal_info_s> m_signal_available;
```

I don't see the need to make ```m_signal_info_s``` a shared_ptr.  It is never passed outside this class.

Take a look at how ```m_signal_available``` is defined and used in the ```MSRIOGroup```.

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+            };
+
+            void parse_config_json(const std::string &config);
+            static json11::Json construct_config_json_obj(
+                const std::string &config);
```
If you've going to have lines earlier in the file that go beyond column ~75, please keep that style consistent throughout the file.  I'd still prefer to see a line break if you get up to column 100 or so.

```suggestion
            static json11::Json construct_config_json_obj(const std::string &config);
```

avilcheslopez replied:
Yeah, let me see if I can clean it up for consistency.

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+                const std::string &config);
+            static void check_json_root(const json11::Json &root);
+            static void check_json_signal(
+                const json11::Json::object::value_type &signal_obj);
```
```suggestion
            static void check_json_signal(const json11::Json::object::value_type &signal_obj);
```

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+                const json11::Json::object::value_type &signal_obj);
+            static void check_json_signal_properties(
+                const std::string &signal_name,
+                const json11::Json::object& properties);
```
```suggestion
            static void check_json_signal_properties(const std::string &signal_name,
                                                     const json11::Json::object& properties);
```

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,174 @@
+            static const std::string M_DEFAULT_CONFIG_FILE_PATH;
+
+            std::map<std::string, std::shared_ptr<m_signal_info_s> >
+                m_signal_available;
```
```suggestion
            std::map<std::string, std::shared_ptr<m_signal_info_s> > m_signal_available;
```

avilcheslopez committed:
Remove check for duplicate signal names in ConstConfigIOGroup

This is due to how json11 parses JSON input: when encountering multiple
entries, it will take the last one encountered.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

Krebrov001 committed:
Changes addressing Brad's review comments

address all comments, **except**:
 - the documentation change in `geopm_pio_const_config.7.rst`, and the removal of the `m_signal_info_s()` constructor

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

avilcheslopez committed:
Improve ConstConfigIOGroup doc, do minor code clean-up

Minor improvements to documentation to clarify how configuration file
can be provided.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Refactor ConstConfigIOGroup constructors

Refactor the specialized constructor to take file paths and have the
default constructor invoke it with defaults. Update the tests accordingly.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add unit tests for ConstConfigIOGroup constructor

Added unit tests to cover cases that require the IOGroup to fall back
to opening the default config.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+            try {
+                config_json = geopm::read_file(user_file_path);
+                load_default = false;
+            } catch (...) {
```
Catch the specific type of exception you expect here: ```const geopm::Exception &ex```

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+        double result = NAN;
+        
+        if (batch_idx < 0 ||
+            static_cast<std::size_t>(batch_idx) >= m_pushed_signals.size()) {
```
```suggestion
            batch_idx >= static_cast<int>(m_pushed_signals.size())) {
```
AFAICT, all the other IOGroups do it this way.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_idx < 0 ||
+            static_cast<std::size_t>(domain_idx) >= it->second->values.size()) {
```
```suggestion
            batch_idx >= static_cast<int>(it->second->values.size()) {
```

bgeltz replied:
AFIACT the other IOGroups all do it this way.

avilcheslopez replied:
Is there a reason why this would be preferred? I'm thinking casting ``int`` to ``unsigned int`` (after we've ensured it's >= 0) is safer than casting ``unsigned int`` to ``int``. I highly doubt the ``values`` vector would ever be so large for that to matter (under normal use cases), but I would still prefer to do the safer cast (in case of malicious input).

bgeltz replied:
Casting a signed int to an unsigned int can result in weird values if the initial value was negative (e.g. 2^32 - 1 or similar for an initial value of -1).  It's why we avoid this sort of thing in other spots in the code.

Casting an unsigned int to a signed int should always be safe unless there's a difference in precision that results in data truncation.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_idx < 0 ||
+            static_cast<std::size_t>(domain_idx) >= it->second->values.size()) {
```
```suggestion
            domain_idx >= static_cast<int>(it->second->values.size()) {
```

bgeltz replied:
Also, this line should be indented a bit more, 1 space beyond the open paren on the line above.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    std::string ConstConfigIOGroup::signal_description(
+        const std::string &signal_name) const
```
```suggestion
    std::string ConstConfigIOGroup::signal_description(const std::string &signal_name) const
```

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+        }
+    }
+
+    Json ConstConfigIOGroup::construct_config_json_obj(
+        const std::string &config)
+    {
+        std::string err;
+        Json root = Json::parse(config, err);
+        if (!err.empty()) {
+            throw Exception("ConstConfigIOGroup::parse_config_json(): "
+                            "detected a malformed JSON string",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return root;
+    }
+
+    void ConstConfigIOGroup::check_json_root(const Json &root)
+    {
+        if (!root.is_object()) {
+            throw Exception("ConstConfigIOGroup::parse_config_json(): "
+                            "the root must be an object",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+    }
```
I don't see the benefit of having these methods separated out from ```parse_config_json()```.  Take a look at some of the other places in the code base we parse json:
    ```FilePolicy::parse_json()```
    ```SaveControlImp::settings()```

Those both have code of the form:
```if (!err.empty() || !root.is_object()) {```

If you agree, please combine all of these error checks into a more condensed form, and put the code in ```parse_config_json()```.

avilcheslopez replied:
I wanted to separate the error checking from the actual parsing to make the parsing code easier to read (by removing the clutter of all the error checking, and keeping the function short and focused on the parsing).

I originally looked at MSRIOGroup's parse method and had a bit of a hard time following it because it's somewhat lengthy, and has some clutter as a result of the checks. A bit more distracting is that it has some of the checks inline while it farms out some checks to other functions.

I looked at FilePolicy's parse method and I do see it has its error checking inline. However, its parsing logic is simple enough that the method is pretty short and easy to read, in which case the error checking doesn't cause a distraction.

I generally try to keep functions reasonably short and focused on one purpose to keep them easy to read, but I know some of this is subjective. In this case I would prefer to keep the parse method as-is, but if you feel strongly that it's a deviation on an important pattern being followed in the codebase and feel it makes the function hard to read, then I'm happy to make the suggested change.

avilcheslopez replied:
Update: Daniel has suggested looking into using json11's ``has_shape()`` method to simplify the checking. Looking at it, I see I can leverage it to at least check required fields and their types. I'll do a pass at improving the checking with that.

avilcheslopez replied:
Update: simplified the parsing method to an extent by reducing the number of "check" functions by leveraging the ``has_shape()`` method.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,451 @@
+        }
+    }
+
+    void ConstConfigIOGroup::check_json_signal_properties(
+        const std::string &signal_name,
+        const Json::object& properties)
```
```suggestion
    void ConstConfigIOGroup::check_json_signal_properties(const std::string &signal_name,
                                                          const Json::object &properties)
```

bgeltz commented on service/src/ConstConfigIOGroup.hpp:
```diff
@@ -0,0 +1,115 @@
+            };
+
+            void parse_config_json(const std::string &config);
+            static json11::Json construct_config_json_obj(const std::string &config);
+            static void check_json_root(const json11::Json &root);
+            static void check_json_signal(const json11::Json::object::value_type &signal_obj);
+            static void check_json_signal_properties(const std::string &signal_name,
+                                                     const json11::Json::object& properties);
+            static void check_json_array_value(const json11::Json& val);
```
Let's talk about what all we really need to expose as a static function at this time.

avilcheslopez replied:
What do you mean? These functions are private and not being exposed.

bgeltz commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,397 @@
+            try {
+                config_json = geopm::read_file(user_file_path);
+                load_default = false;
+            } catch (const Exception &ex) {
```
```suggestion
            }
            catch (const geopm::Exception &ex) {
```

bgeltz reviewed (changes requested):
I'll take a better look on Monday

bgeltz commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,658 @@
+};
+
+const std::string ConstConfigIOGroupTest::M_CONFIG_FILE_PATH =
+        "/tmp/const_config_test.json";
```
```suggestion
        "const_config_test.json";
```
Is there a need to write these to tmp as opposed to the CWD?  If we put them in tmp, we may have other GEOPM users who are also writing to the same spot.  If 2 people are running make check at the same time, this is a problem.

avilcheslopez replied:
Thanks, will fix.

bgeltz commented on service/test/ConstConfigIOGroupTest.cpp:
```diff
@@ -0,0 +1,658 @@
+    EXPECT_EQ(iogroup.sample(1), 1700);
+    EXPECT_EQ(iogroup.read_signal("CONST_CONFIG::CPU_CORE_FREQUENCY", GEOPM_DOMAIN_CPU, 2), 1070);
+    EXPECT_EQ(iogroup.read_signal("CONST_CONFIG::GPU_CORE_FREQUENCY", GEOPM_DOMAIN_GPU, 0), 1500);
+    {
+        auto expected_function = geopm::Agg::average;
+        auto actual_function = *iogroup.agg_function("CONST_CONFIG::CPU_CORE_FREQUENCY")
+            .target<double (*) (const std::vector<double>&)>();
+        EXPECT_EQ(expected_function, actual_function);
+    }
```
Take a look at ```service/test/geopm_test.hpp```.  It has helper functions for this purpose so you can write this in the form:
```
std::function<double(const std::vector<double> &)> func = iogroup.agg_function("CONST:CONFIG::CPU_CORE_FREQUENCY");
// You could also use: auto func = iogroup.agg_function("CONST:CONFIG::CPU_CORE_FREQUENCY");
EXPECT_TRUE(is_agg_average(func));
```

Please write these all this way.

avilcheslopez replied:
Ah, cool. Thanks!

avilcheslopez committed:
ConstConfigIOGroup code clean-up: formatting changes

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
ConstConfigIOGroupTest clean-up: minor refactoring

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Refactor ConstConfigIOGroup JSON parsing method

Consolidate JSON checking helper functions

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/src/ConstConfigIOGroup.cpp:
```diff
@@ -0,0 +1,435 @@
+                }
+                values.push_back(val.number_value());
+            }
+
```
Add check that length of array matches topo.nun_domain() for the system.  This will avoid a late failure that would occur when a user tries to read the value.
